### PR TITLE
JAVA-2611: Spec updates

### PR DESCRIPTION
**Pull request by jyemin**
Restrict use of TXT records to a single record containing only authSource and/or replicaSet options,
and default ssl to true for mongodb+srv connection strings.

The base commit is https://github.com/jyemin/mongo-java-driver/pull/78.

Patch build: https://evergreen.mongodb.com/version/5a16fa94e3c331755d0003b5

### CODE REVIEW

File: driver-async/src/test/functional/com/mongodb/async/client/InitialDnsSeedlistDiscoveryTest.java [hunk 1]
```diff
@@ -96,48 +94,35 @@ public void shouldResolve() {
...
-                    assertEquals(entry.getValue().asNumber().intValue(), (int) connectionString.getSocketTimeout());
-                } else if (entry.getKey().equals("readPreference")) {
-                    assertEquals(entry.getValue().asString().getValue(), connectionString.getReadPreference().getName());
-                } else if (entry.getKey().equals("readPreferenceTags")) {
-                    assertEquals(asTagSetList(entry.getValue().asArray()),
-                            ((TaggableReadPreference) connectionString.getReadPreference()).getTagSetList());
+                } else if (entry.getKey().equals("ssl")) {
+                    assertEquals(entry.getValue().asBoolean().getValue(), connectionString.getSslEnabled());
+                } else if (entry.getKey().equals("authSource")) {
+                    // ignoring authSource for now, because without at least a userName also in the connection string,
```
Comment (jyemin) [inline driver-async/src/test/functional/com/mongodb/async/client/InitialDnsSeedlistDiscoveryTest.java]: This is unfortunate but would require more surgery in ConnectionString than I want to take on right now in order to be able to make an assertion here.  So instead testing it on ConnectionStringSpecification

File: driver-async/src/test/functional/com/mongodb/async/client/InitialDnsSeedlistDiscoveryTest.java [hunk 2]
```diff
@@ -96,48 +94,35 @@ public void shouldResolve() {
...
         }
-        assumeTrue(isDiscoverableReplicaSet());
-
         final CountDownLatch latch = new CountDownLatch(1);
 
         ConnectionString connectionString = new ConnectionString(uri);
 
+        SslSettings sslSettings = getSslSettings(connectionString);
+
+        assumeTrue(isDiscoverableReplicaSet() && getSslSettings().isEnabled() == sslSettings.isEnabled());
```
Comment (jyemin) [inline driver-async/src/test/functional/com/mongodb/async/client/InitialDnsSeedlistDiscoveryTest.java]: The end result is that the test URIs which resolve to ssl=true will be tested here when running against an SSL-enable replica set, and the ones which resolve to ssl=false will be tested here when running against an SSL-disable replica set.

File: driver-core/src/main/com/mongodb/ConnectionString.java [hunk 3]
```diff
@@ -455,10 +463,10 @@ private void translateOptions(final Map<String, List<String>> optionsMap) {
                 connectTimeout = parseInteger(value, "connecttimeoutms");
             } else if (key.equals("sockettimeoutms")) {
                 socketTimeout = parseInteger(value, "sockettimeoutms");
-            } else if (key.equals("sslinvalidhostnameallowed") && parseBoolean(value, "sslinvalidhostnameallowed")) {
-                sslInvalidHostnameAllowed = true;
-            } else if (key.equals("ssl") && parseBoolean(value, "ssl")) {
-                sslEnabled = true;
+            } else if (key.equals("sslinvalidhostnameallowed")) {
+                sslInvalidHostnameAllowed = parseBoolean(value, "sslinvalidhostnameallowed");
```
Comment (jyemin) [inline driver-core/src/main/com/mongodb/ConnectionString.java]: The change for the boolean params allow tests to assert that the values of the these options were set even when they are set to their default values.

File: driver-core/src/main/com/mongodb/internal/dns/DnsResolver.java [hunk 4]
```diff
@@ -100,19 +100,18 @@ public static String resolveAdditionalQueryParametersFromTxtRecords(final String
             Attributes attributes = dirContext.getAttributes(host, new String[]{"TXT"});
             Attribute attribute = attributes.get("TXT");
             if (attribute != null) {
-                StringBuilder additionalQueryParametersBuilder = new StringBuilder();
                 NamingEnumeration<?> txtRecordEnumeration = attribute.getAll();
-                while (txtRecordEnumeration.hasMore()) {
+                if (txtRecordEnumeration.hasMore()) {
```
Comment (jyemin) [inline driver-core/src/main/com/mongodb/internal/dns/DnsResolver.java]: Only allow a single TXT record

File: driver-async/src/test/functional/com/mongodb/async/client/InitialDnsSeedlistDiscoveryTest.java (hunk 1 above)
Comment (rozza) [reply driver-async/src/test/functional/com/mongodb/async/client/InitialDnsSeedlistDiscoveryTest.java]: Ack

File: driver-core/src/main/com/mongodb/ConnectionString.java (hunk 3 above)
Comment (rozza) [reply driver-core/src/main/com/mongodb/ConnectionString.java]: Ack

File: driver-core/src/test/resources/initial-dns-seedlist-discovery/README.rst [hunk 5]
```diff
@@ -9,7 +9,8 @@ Test Setup
 ----------
 
 Start a three-node replica set on localhost, on ports 27017, 27018, and 27019,
-with replica set name "repl0".
+with replica set name "repl0". The replica set MUST be started with SSL
```
Comment (rozza) [inline driver-core/src/test/resources/initial-dns-seedlist-discovery/README.rst]: Do we need this file in the resources?

Comment (jyemin) [reply driver-core/src/test/resources/initial-dns-seedlist-discovery/README.rst]: It's not essential but it does explain what's going on.  Without knowing what's in the SRV/TXT DNS records the tests don't make sense.

**Pull request closed without merging**